crates/sozo/ops/src/migration/mod.rs

```diff
@@ -772,3 +775,129 @@ where
+
+    let InvokeTransactionResult { transaction_hash } = WorldContract::new(world_address, migrator)
+        .set_metadata(&resource_metadata)
+        .send()
+        .await
+        .map_err(|e| {
+            ui.verbose(format!("{e:?}"));
+            anyhow!("Failed to set resource metadata: {e}")
+        })?;
+
+    TransactionWaiter::new(transaction_hash, migrator.provider()).await?;
+
+    ui.print_sub(format!("Set Metadata transaction: {:#x}", transaction_hash));
+    ui.print_sub(format!("Metadata uri: ipfs://{hash}"));
+
+    Ok(())
+}
+
+async fn upload_model_metadata<P, S>(
+    ui: &Ui,
+    migrator: &SingleOwnerAccount<P, S>,
+    world_address: FieldElement,
+    model_names: Vec<String>,
+    metadata: &HashMap<String, ArtifactMetadata>,
+) -> Result<()>
+where
+    P: Provider + Sync + Send + 'static,
+    S: Signer + Sync + Send + 'static,
+{
+    for model_name in model_names {
+        if let Some(m) = metadata.get(&model_name) {
+            ui.print(italic_message(&model_name).to_string());
+            match m.upload().await {
+                Ok(hash) => {
+                    let resource_id =
+                        get_selector_from_name(model_name.as_str()).expect("ASCII string");
+                    register_resource_metadata(ui, migrator, world_address, hash, resource_id)
+                        .await?;
+                }
+                Err(err) => {
+                    ui.print_sub(format!("Failed to set resource metadata:\n{err}"));
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn upload_contract_metadata<P, S>(
+    ui: &Ui,
+    migrator: &SingleOwnerAccount<P, S>,
+    world_address: FieldElement,
+    contract_data: Vec<(String, FieldElement)>,
+    metadata: &HashMap<String, ArtifactMetadata>,
+) -> Result<()>
+where
+    P: Provider + Sync + Send + 'static,
+    S: Signer + Sync + Send + 'static,
+{
+    for (contract_name, contract_address) in contract_data {
+        if let Some(m) = metadata.get(&contract_name) {
+            ui.print(italic_message(&contract_name).to_string());
+            match m.upload().await {
+                Ok(hash) => {
+                    register_resource_metadata(ui, migrator, world_address, hash, contract_address)
+                        .await?;
+                }
+                Err(err) => {
+                    ui.print_sub(format!("Failed to set resource metadata:\n{err}"));
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn upload_metadata<P, S>(
+    ui: &Ui,
+    migrator: &SingleOwnerAccount<P, S>,
+    world_address: FieldElement,
+    model_names: Vec<String>,
+    contract_data: Vec<(String, FieldElement)>,
+    metadata: &HashMap<String, ArtifactMetadata>,
+) -> Result<()>
+where
+    P: Provider + Sync + Send + 'static,
+    S: Signer + Sync + Send + 'static,
+{
+    ui.print(" ");
+    ui.print_step(5, "🌐", "Uploading metadata...");
```

Comment:
As this step may be skipped when sozo is used with `--offline`, we eventually want the manifest update to be `5` or `6` depending on the case.